### DISCUSSION TITLE
PHP Performance Monitoring / Tracing support

### DISCUSSION THREAD

**Discussion by stayallive** [stayallive/wp-sentry#113] [Ideas]
As originally request by @archon810 on getsentry/sentry-php#1062.

---

Hi,

This is a follow-up to https://github.com/getsentry/sentry-laravel/pull/358#issuecomment-669804042, with the goal of adding the new PHP Tracing/Performance support specifically for WordPress.

Thanks.

**Comment by archon810** [comment]
Thanks, didn't even realize this project existed. Looks unofficial but semi-official, considering your affiliation with Sentry.

I hope Perf can become part of this project when it's ready.

**Comment by stayallive** [comment]
This is the best place for it since the official PHP SDK will never support WordPress since it's just for the PHP integration and I'm not sure there is interest in officially supporting WordPress like Laravel or Symfony. The issue in the PHP SDK would therefore not do much good 👍 

Assuming you are already using Sentry in some capacity in WordPress using this plugin or not. Once the tracing code becomes available in the Sentry PHP SDK this plugin or your custom code can integrate it with WordPress to bring tracing to WordPress. If you are using another WordPress plugin that provides the Sentry integration you might want to also open up an issue there to let them know you would be interested in this.

I might implement this myself or maybe someone is willing to submit a PR for it, we'll see but it's good to have an issue open to see if there's interest in supporting it 😄 

(sorry my finger slipped the close button trying to place this comment)

**Comment by archon810** [comment]
We're just using the PHP SDK and I wasn't aware of this project until you moved this issue into it.

Looking at the benefits vs running our own code, we may end up switching, especially if Performance is added in.

**Comment by stayallive** [comment]
Yeah sorry about that, I though somehow you were using this plugin, my bad for assuming!

The performance code will be added to the PHP SDK in the future so you could implement it yourself just like you currently implemented the PHP SDK in your application

However there will be no framework/WordPress specific code in the PHP SDK itself for performance monitoring so that has to be added by the developer. The PHP SDK by itself will also not show in-depth performance information out of the box because it simply cannot, this is to the developer and or integration developer to implement (the Laravel one is a proof of concept and a testing ground).

Once it's stable enough in the PHP SDK for me to start including here I will see if I can provide some hooks in WordPress, I know a few users who also would like to see performance metrics for their WordPress application so I'm sure it will happen at some point or another 😄

**Comment by archon810** [comment]
3.0 is out https://github.com/getsentry/sentry-php/releases/tag/3.0.0, what do you think about kicking things off @stayallive?

**Comment by stayallive** [comment]
I don't have the time for this right as this moment, so feel free to kick things off 😉 

Even if you would upgrade to the PHP SDK 3.0 performance monitoring would not magically work and still require quite some work to get working and add all the required tracing bits and pieces 😄 

But upgrading to the PHP SDK version 3.0 would be step 1 in the process.

**Comment by stayallive** [comment]
For the followers of this issue, I have just released 4.0.0 which contains the 3.x PHP SDK and thus performance monitoring support.

This however _DOES NOT_ mean that performance tracing is working with a flick of a config switch yet 😉  

But with some manual labour you could get it working now because the SDK is up-to-date.

Keeping this issue open to track progress on integrating in WordPress and adding automatic performance tracing.

**Comment by Trisky** [comment]
Hi! I needed the performance feature so I  implemented it in the mu-plugin. I Did it this way because I keep that in version control and install wp-sentry through wpackagist.

It kinda works but it **doesn't show the stack trace** in sentry performance and I'm kinda lost on what I'm missing

```php
$wp_sentry = __DIR__ . '/../plugins/wp-sentry-integration/wp-sentry.php';

if (!file_exists($wp_sentry)) {
    trigger_error('Sentry is not being mu-loaded because the plugin files are not there');
    return;
}

require $wp_sentry;

define( 'WP_SENTRY_MU_LOADED', true );

/*
 * Sentry performance activation for wordpress
 */
$sentryDsn = defined('WP_SENTRY_DSN') && (WP_SENTRY_DSN !== false);
$sentryTracingRate = defined('WP_SENTRY_TRACES_SAMPLE_RATE') ? WP_SENTRY_TRACES_SAMPLE_RATE : false;
if ($sentryDsn && is_numeric($sentryTracingRate) && $sentryTracingRate != 0) {
    /*
     * First we add the sample rate and set the current hub so the new transaction is attached to it.
     */
    wp_sentry_safe(function (\Sentry\State\HubInterface $sentryHub) {
        $rate = WP_SENTRY_TRACES_SAMPLE_RATE ?? 0;
        $sentryHub->getClient()->getOptions()->setTracesSampleRate($rate);
        \Sentry\SentrySdk::setCurrentHub($sentryHub);
    });
    $transactionContext = new \Sentry\Tracing\TransactionContext();
    $transactionContext->setName($_SERVER['REQUEST_URI']);
    $transactionContext->setOp('http.server');
    $transactionContext->setData([
        'url' => $_SERVER['REQUEST_URI'] ?? 'no_url',
        'method' => strtoupper($_SERVER['REQUEST_METHOD'] ?? '')
    ]);
    $transactionContext->setStartTimestamp($_SERVER['REQUEST_TIME_FLOAT'] ?? null);
    $transaction = \Sentry\startTransaction($transactionContext);
    /*
    * We create a span for our transaction (is this ok?)
    */
    $spanContext = new \Sentry\Tracing\SpanContext();
    $spanContext->setOp('wordpress');
    $GLOBALS['SENTRY_PERFORMANCE_TRANSACTION'] = $transaction;
    $GLOBALS['SENTRY_PERFORMANCE_SPAN'] = $transaction->startChild($spanContext);
  
    /*
    * On wp shutdown, we finalize the transaction.
    */
    add_action('shutdown', function () {
        $sentryTransaction = $GLOBALS['SENTRY_PERFORMANCE_TRANSACTION'] ?? null;
        $sentrySpan = $GLOBALS['SENTRY_PERFORMANCE_SPAN'] ?? null;
        if ($sentryTransaction instanceof \Sentry\Tracing\Transaction && $sentrySpan instanceof \Sentry\Tracing\Span) {
            $sentryTransaction->finish();
            $sentrySpan->finish();
        } else {
            trigger_error('SENTRY global variables have been modified');
        }
    });
}
```

Do you guys have an idea on why is the stack trace not working? 🤔   The stacktrace does work for JS but not for PHP as you can see here:



![image](https://user-images.githubusercontent.com/8531036/112603346-2b7cf900-8e15-11eb-9c92-7eee57af806d.png)


When I manage to get the stack trace running I will submit a proper PR for it (not a mu-plugin hack haha).

**Comment by stayallive** [comment]
I'm not able to dig in but I hope these links help you in the right direction:

Starting the transaction:
https://github.com/getsentry/sentry-laravel/blob/master/src/Sentry/Laravel/Tracing/Middleware.php#L80-L118

Finishing the transaction:
https://github.com/getsentry/sentry-laravel/blob/master/src/Sentry/Laravel/Tracing/Middleware.php#L76

Adding a child to the transaction:
https://github.com/getsentry/sentry-laravel/blob/master/src/Sentry/Laravel/Tracing/Middleware.php#L104-L108

Looking at your code I would only expect 2 things in the stack, the `http.server` and `wordpress` below that, but I think the reason it's not working because you are finishing the span _after_ the transaction, so the transaction is already sent when you end the span, try switching that around.

Finishing the transaction sends the transaction so that should be the very last thing you do.

Hope this helps a bit!

**Comment by Trisky** [comment]
Oh I  didn't think about finishing the span before the transaction haha. Great.

I've been also checking the implementation for Laravel for inspiration so I'll improve the code a bit removing the global variable mess I made.

I'll try to make one span for the plugins and an other for the theme as well. What do you think?

Thanks. 🎉

**Comment by stayallive** [comment]
We probably want to end up trying to do a bit more like listening for `plugins_loaded`, `setup_theme` and `wp_loaded` and recording those times as spans too since I think that might be interesting data.

And maybe we can use other hooks to measure HTTP requests and possibly even SQL queries.

But this is already off to a great start 👍

**Comment by Trisky** [comment]
I've modified the mu-plugin adding this code after `define( 'WP_SENTRY_MU_LOADED', true );` to add sentry-performance but I didn't get too far away. It tells me which endpoints are running slow (good) but I didn't  figure out how to activate the stack analysis that works for Python, javascript or Laravel itself. So it only shows this in the event details:



![image](https://user-images.githubusercontent.com/8531036/115214921-46384a00-a103-11eb-93a4-3f9761287f3e.png)

```php
<?php

/*
 * Sentry performance activation for wordpress
 */
$sentryDsn = defined('WP_SENTRY_DSN') && (WP_SENTRY_DSN !== false);
$sentryTracingRate = defined('WP_SENTRY_TRACES_SAMPLE_RATE') ? WP_SENTRY_TRACES_SAMPLE_RATE : false;
if ($sentryDsn && is_numeric($sentryTracingRate) && $sentryTracingRate != 0) {
    # 1 - set up tracing in our hub
    wp_sentry_safe(function (\Sentry\State\HubInterface $sentryHub) {
        $rate = WP_SENTRY_TRACES_SAMPLE_RATE ?? 0;
        $sentryHub->getClient()->getOptions()->setTracesSampleRate($rate);
        \Sentry\SentrySdk::setCurrentHub($sentryHub);
    });
    # 2 set up transaction context
    $transactionContext = new \Sentry\Tracing\TransactionContext();
    if($_SERVER['REQUEST_URI']?? null){
        $name = "$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
    }else{
        if ( defined( 'DOING_CRON' ) ) {
            $name = 'cron';
        }else{
            $name = 'wp-cli';
        }
    }
    # 3 set up transaction context
    $transactionContext->setName($name);
    $transactionContext->setOp('http.server');
    $transactionContext->setData([
        'url' => $name,
        'method' => strtoupper($_SERVER['REQUEST_METHOD'] ?? '')
    ]);
    $transactionContext->setStartTimestamp($_SERVER['REQUEST_TIME_FLOAT'] ?? microtime(true) );

    # 4 start transaction
    $transaction = \Sentry\SentrySdk::getCurrentHub()->startTransaction($transactionContext);
    \Sentry\SentrySdk::getCurrentHub()->setSpan($transaction);

    # 5 set up span context
    $spanContext = new \Sentry\Tracing\SpanContext();
    $spanContext->setOp('wordpress_span');
    $spanContext->setStartTimestamp(microtime(true));

    # 6 start span
    $appSpan = $transaction->startChild($spanContext);

    \Sentry\SentrySdk::getCurrentHub()->setSpan($appSpan);

    add_action('shutdown', function () {
        # 7 finish span
        $sentrySpan = \Sentry\SentrySdk::getCurrentHub()->getSpan();
        if($sentrySpan){
            $sentrySpan->finish();
        }
        $sentryTransaction = \Sentry\SentrySdk::getCurrentHub()->getTransaction();
        # 8 finish transaction
        if($sentryTransaction){
            \Sentry\SentrySdk::getCurrentHub()->setSpan($sentryTransaction);
            $transaction = \Sentry\SentrySdk::getCurrentHub()->getSpan();
            $transaction->finish();
        }
    });
}


```
Should this actually work and I'm doing something wrong?  Or am I missing to implement something?

**Reply by mootensai** [reply]
hi @Trisky, thank you for this script! I tried this, but on the number 8 finish transaction, the 
```
$sentryTransaction = \Sentry\SentrySdk::getCurrentHub()->getTransaction();
```
always return null. do you know the cause of this? 

strangely, I happen to make it work once.

**Comment by codeagencybe** [comment]
@stayallive 
Is this feature already (official) available in the current version/implementation? 
I was just reading the plugin docs and wp-config params and it only mentions about javascript performance. 
Does this mean php performance is not there yet? 

Im also looking for performance monitoring and already using Sentry. This new feature would be amazing to finally understand performance better from Sentry. 

Thanks!

**Comment by stayallive** [comment]
Hi @codeagencybe, there are 2 parts to performance tracing.

- [x] The PHP SDK to have the code needed to hold and send the traces to Sentry Server
- [ ] The code needed in the WordPress plugin to actually created traces and profile certain WordPress features

Unfortunately I have not had the time to put into this so it will either require some community effort or some patience before this becomes fully available and usable in this WordPress plugin.

There has been some progress by the community (https://github.com/stayallive/wp-sentry/issues/68#issuecomment-822337859) but no complete pieces of code as of yet to just drop-in and have some useful performance tracing.

To give a bit of context, there is no magic `enabled_tracing()` function or `trace_this()` method in PHP unfortunately so all tracing will have to implemented manually as the Sentry team has done for various other languages and also for Symfony and Laravel, but there is no easy way to enable tracing without platform/framework specific code and that why it's taking a while.

**Comment by codeagencybe** [comment]
@stayallive 

Thanks for the feedback! 
By all means, I'm not putting any stress on you or anybody on this. 
I just by accident discovered this thread from a Google search and read there was some work done but didn't understand the full scope yet what is done and what not (yet). 

I can totally understand it takes a lot effort to implement this for every possible language in the market. 

As a side note/idea; 
There is a new relic plugin for Wordpress, would that plugin help as a concept for this Sentry plugin? 
At least it could help how they do the "magic"? 

Thanks again!

**Comment by stayallive** [comment]
@codeagencybe, no worries, it was a good question and good time to explain it fully what the status is and what would be needed to get this off the ground (also for future Googlers ending up here). 

Also good call about the New Relic plugin. I can only find an [unofficial](https://nl.wordpress.org/plugins/wp-newrelic/) one by 10up which adds metadata to the New Relic traces collected by their PHP agent and does not actually do much (if any) tracing itself.

There is an important difference between how New Relic and Sentry does performance tracing. New Relic uses a PHP extension to directly hook into the PHP engine which does allow them to track way more much more easier but also as a downside requires the installation of a PHP extension which is not always possible. As far as I know there is no intention as of this moment to also built a Sentry PHP extension.

So there is no code from New Relic we could take inspiration from in this case unfortunately.

This is also the same for Datadog which also offer PHP APM with WordPress integration from the looks of their docs, again with a PHP extension as the backbone of the actual tracing.

**Comment by codeagencybe** [comment]
@stayallive 

Yep that's the one plugin from 10up I also found. 
I didn't know there was a php extension involved, it looked like it was just a plugin thats pushing data to new relic service, and from there it's on NR end to handle it down the road. 
Installing an extension for me is no problem as we run and manage our own servers, but I can understand this is not a solution for everyone on lockdowned machines and/or limited access to manage their hosting environment. 

I have never used Datadog before, just know them from name, so I can't tell or vouch anything on that one. 
As a side idea, would lifting on that existing php extension be an option instead of developing a Sentry version (if ever needed)? 
I have no experience to claim anything be compatible or anything, just thinking out loud ideas and expressions. 
Sometimes its easier to use what's already in the market instead of re-inventing the wheel over and over. 
Same story for Glitchtip it seems who forked from Sentry but still rely on the same plugins and base. 
Not good for longterm, but it gives opportunity to get something started and working and then build a seperate solution (that's hopefully better).

**Comment by codeagencybe** [comment]
@Trisky 

Did you manage to get it working? Do you have the code for your mu-plugin somewhere so I can give it a try also? 

Thanks!

**Comment by Trisky** [comment]
@codeagencybe 
Hi Fabio, you can find what I have [here in my comment](https://github.com/stayallive/wp-sentry/issues/68#issuecomment-822337859). I didn't dive into this issue again.
Best,
Sebastián.

**Comment by smeubank** [comment]
if someone is using [Laravel with WP ](https://wpengine.com/resources/wordpress-laravel/)- would it be possible to take advantage of the laravel auto instrumentation?

**Comment by Trisky** [comment]
@smeubank 
Hi!
I'm not using it, at first glance it should work but only for what happens inside the Laravel plugin... What happens before and after it's loaded would remain in the dark I guess.
https://codex.wordpress.org/Plugin_API/Action_Reference

**Comment by stayallive** [comment]
Yeah the Laravel / WordPress hybrids will not have much benefits from the Laravel instrumentation only the Laravel part will be instrumented.

(I wasn't even aware that that was a thing 😅 people think of some crazy concepts)

**Comment by robertoandrade** [comment]
Thanks a lot for sharing this! Regarding the intermediary spans to identify which hooks or db queries are slow, I wonder if one could reuse the code from the WP CLI "profile"plugin: https://github.com/wp-cli/profile-command

**Comment by ShahinSorkh** [comment]
Do you have any updates on this discussion maybe? hopefully? possibly?

**Comment by joshfester** [comment]
Hi @stayallive ,

I've been bikeshedding all my monitoring possibilities with Wordpress for the last week and feel like Sentry is probably going to be the easiest way to track errors/performance. I came across a [wordpress implementation](https://github.com/opentelemetry-php/contrib-auto-wordpress/blob/main/src/WordpressInstrumentation.php) by the OpenTelemetry team. It seems like a good example of how to instrument tracing in wordpress, so I thought I'd share here in case it's of any help to you. I may take a stab at a pull request myself but only if it ends up being fairly easy as I don't have a lot of time to dedicate to it.

**Reply by stayallive** [reply]
Thanks for sharing that. Keep in mind that that library requires a separate PHP extension to be installed to work while Sentry doesn't have that requirement (except for actual [profiling](https://docs.sentry.io/platforms/php/profiling/) which already works, but we use an existing solution for that). 

So their code cannot work 1:1 since they have a lot more control and tools than we have. We have to do it in userland code, which is not impossible just requires a little more thought and carefulness when implementing (and possibly cooperation from the code we are trying to trace) so although their implementation is good to look at it is not that useful here unfortunately.

**Reply by joshfester** [reply]
Ah ok that makes sense. It's crazy to me that there's no official Wordpress integrations for tools like Sentry. You'd think there would be a huge market for it

**Reply by stayallive** [reply]
I also thought there would be quite a bigger market but if you compare downloads/installations to PHP frameworks for example the usage is still not that significant 🥲 

Either WordPress folk don't care that much or there are good-enough solutions already. Also a bit of a chicken and the egg problem because there are no (except for profiling) performance features in the plugin which might (or might not) make this plugin more useful and increase numbers.

It's also tricky to invest much time since that would be unpaid for contributors at the moment. Still planning on working on it myself (and probably others have similar plans too) but finding the free time for this is not trivial and Sentry (speaking unofficially) might also not see the return on an investment at this point. 

It will come, it will just take time 😎

**Reply by joshfester** [reply]
I hope that it will gain some momentum! I will probably be putting some time into it at some point as I work on a large Wordpress project and I consider performance tracking to be a core requirement. 

This plugin seems like the best starting point. All the other open-source solutions seem to be using OpenTelemetry behind the scenes, and they seem very complicated compared to Sentry. Are you aware of any platforms that come close and don't have vendor lock-in like NewRelic?

**Reply by stayallive** [reply]
OpenTelemetry (especially in the PHP space) seems to make everything way more complicated. I believe there was some work for Sentry to be more OpenTelemetry compatible (either on the ingest and the SDK side).

The biggest power for Sentry is that there are no PHP plugins needed which all other vendors require that I know of. And although this isn't perfect and we can't measure all the things we would want but we get really useful data like already done for Laravel and Symfony add to that Sentry server is self-hosted and the SDKs are open source. I don't know of any competitor coming close at the moment (especially in the PHP space). 

(full disclosure in case you haven't noticed on my profile, I am a freelance contributor to the official PHP SDKs at Sentry next to building the WordPress version in my own time)

**Reply by joshfester** [reply]
Ah I see. Thanks for the info! I did notice your contributions, that is a big reason why I felt good about using this plugin 😁

**Comment by x-xyzxyz** [comment]
Hello,
does anyone know what's the right way to have PHP Performance and/or Profiling work with the current state of `wp-sentry`?

What I've done:

- installed `php-excimer` (which I confirmed to be working fine)
- added this to `functions.php`:
```php
	add_filter(
		'wp_sentry_options',
		function ( \Sentry\Options $options ) {

			$options->setTracesSampleRate( 1.0 );
			$options->setProfilesSampleRate( 1.0 );
			$options->setEnableTracing( true );

			return $options;
		}
	);
```
- added this example from their documentation to my code:
```php
function expensive_operation() {
	echo 'hello';
}

// Setup context for the full transaction
$transactionContext = ( new \Sentry\Tracing\TransactionContext() )
	->setName( 'Example Transaction' )
	->setOp( 'http.server' );

// Start the transaction
$transaction = \Sentry\startTransaction( $transactionContext );

// Set the current transaction as the current span so we can retrieve it later
\Sentry\SentrySdk::getCurrentHub()->setSpan( $transaction );

// Setup the context for the expensive operation span
$spanContext = ( new \Sentry\Tracing\SpanContext() )
	->setOp( 'expensive_operation' );

// Start the span
$span1 = $transaction->startChild( $spanContext );

// Set the current span to the span we just started
\Sentry\SentrySdk::getCurrentHub()->setSpan( $span1 );

// Calling expensive_operation()
expensive_operation();

// Finish the span
$span1->finish();

// Set the current span back to the transaction since we just finished the previous span
\Sentry\SentrySdk::getCurrentHub()->setSpan( $transaction );

// Finish the transaction, this submits the transaction and it's span to Sentry
$transaction->finish();
```

I think this covers everything their setup instructions mention to do, but the Profiling section in my Sentry account is always empty.
How are you all setting this up?

FYI the plugin does send all the other standard errors successfully to my PHP and JS projects, I just can't make it send stuff to Profiling.

Thank you

**Comment by stayallive** [comment]
I have started the work for basic performance profiling capabilities in #190.

If you are feeling really adventurous you should be able to give it a go, but no guarantees (yet).

**Reply by archon810** [reply]
Very exciting, thank you Alex @stayallive!

**Reply by eduwass** [reply]
Awesome, thanks for your work @stayallive !

**Reply by Trisky** [reply]
Hi! It looks promising! Great work!

Some stuff I've added to my setup for sentry-performance which is has some features but also some bugs  

1. `sql.origin` added with `$context->setData()` for each SQL span, I add (a truncated version of) `$query_callstack` to the span so it's easy to pinpoint where the sql request came from:
2. same as `1.` but for `http.client` spans.
3. Add frequent Wordpress actions as spans. Examples: `Init`, `wp_loaded` , `setup_theme` , `do_feed_rss2`
4. Same as `3.` but for some Plugins. We should leave this for a later stage (or never)
6. Improvements for transaction name:`$context->setName()` 
5a.`wp-cli` support: add the correct transaction name 
5b.  `wp-admin/post.php` & `wp-admin/admin-ajax.php` specific naming: name the transaction after the `action` or `post_id` parameter. This would make the name more useful.
6. Create a hook to create custom spans. Useful for theme or plugin developers. Something like this:


```php
do_action('wp_sentry_span_start', 'my_cute_span');
do_heavy_stuff();
do_action('wp_sentry_span_finish');
```


Remarks:

a. You're naming the transaction with just the URI, but for multisite installs it would be useful to have the full url without the get parameters. 
b. You're enabling `SAVEQUERIES`. Obviously we need it to trace sql queries, but a disclaimer about it should be added to the Projects readme? 


@stayallive 
I'll work on a merge request towards `feature/performance` for `1.` and `2.` if you agree.

**Reply by stayallive** [reply]
For 1 and 2 you should add the following to the span data:

```
[
    'code.filepath' => 'wp-content/themes/functions.php', // should be the full path
    'code.function' => 'the_function_or_method_name',
    'code.lineno' => 123,
];
```

This way Sentry can render the origin in the UI, we should _not_ add the (full or shortened) callstack.

3. This can only work if we have a "start" and "end" hook for those actions. Most do not. But I'm certainly open to suggestions.
4. Agreed.
5. I don't want to support CLI stuff at first (we also don't for Laravel/Symfony)
6. That is extremely complicated with `do_action`, we have `\Sentry\trace()` you should use for that purpose

Remarks:

a. Good point on multi-site, however also now out of scope for first version (will re-mark this in the readme)
b. Yes. 

Feel free to try and make 1 & 2 work!

I have updated the PR with the items to-do.

**Reply by Trisky** [reply]
For the actions that don't have "start" and "end", what I did is to hook with priority `0` for start and `PHP_INT_MAX` for end. 
Not the cleanest solution out there, but the one I could come up with. 


Nice tip about the span data, thats way better than my solution. I'll give it a shot!

**Reply by stayallive** [reply]
That's actually not a half bad solution to be honest 😅 We'd have to compile a list with interesting actions, would not mind to do that in a followup update though if it takes some time. 

If you are curious, the query source is rendered like this when available:

![afbeelding](https://github.com/stayallive/wp-sentry/assets/1090754/0fe1c965-23f8-4d72-850b-09e7b812a3cb)

Example from Laravel, but you get the point 😄

**Reply by titouanmathis** [reply]
Tried it on a local environment, the data I get in Sentry is already interesting and will help me identify bottlenecks. Thanks for the work!

**Reply by stayallive** [reply]
That is great to hear, especially that it didn't break immediately 😉 
Thanks for taking the time to try it out 💪

**Reply by stayallive** [reply]
So the query callstack was not what I was expecting and is pretty useless so I opted to add it as a simple string for now and will work on a better implementation later if even possible.

I have updated the PR (#190) and am getting closer to having almost all features complete for a first release. Now would be a great time to jump in and try it out a bit to see how it goes.

**Comment by stayallive** [comment]
Tracing is now released: https://github.com/stayallive/wp-sentry/releases/tag/v7.17.0
